database/pkg.go

```diff
@@ -0,0 +1,219 @@
+package database
+
+import (
+	"encoding/json"
+	"fmt"
+	liberr "github.com/konveyor/controller/pkg/error"
+	"github.com/konveyor/tackle2-hub/api"
+	"github.com/konveyor/tackle2-hub/model"
+	"github.com/konveyor/tackle2-hub/settings"
+	"gorm.io/driver/sqlite"
+	"gorm.io/gorm"
+	"gorm.io/gorm/schema"
+	"io/ioutil"
+	"os"
+	"path"
+	"reflect"
+	"strings"
+)
+
+//
+// DB constants
+const (
+	ConnectionString = "file:%s?_foreign_keys=yes"
+)
+
+//
+// Migration versions
+const (
+	Version210     = "v2.1.0"
+	CurrentVersion = Version210
+)
+
+//
+// Setup the DB and models.
+func Setup() (db *gorm.DB, err error) {
+	db, err = open()
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+
+	var seeded bool
+	seeded, err = isSeeded(db)
+	if err != nil {
+		return
+	}
+
+	dbVersion := &model.Setting{}
+	result := db.FirstOrCreate(&dbVersion, model.Setting{Key: api.HubDBVersion})
+	if result.Error != nil {
+		err = result.Error
+		return
+	}
+
+	// not seeded and empty version record means this is a fresh db
+	// that is already on the latest schema version. Replace empty
+	// version record with current version.
+	if !seeded && dbVersion.Value == nil {
+		err = updateVersion(db, CurrentVersion)
+		if err != nil {
+			return
+		}
+	}
+
+	err = migrate(db)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+
+	if !seeded {
+		err = seed(db)
+		if err != nil {
+			err = liberr.Wrap(err)
+			return
+		}
+	}
+
+	return
+}
+
+//
+// Open and automigrate the DB.
+func open() (db *gorm.DB, err error) {
+	db, err = gorm.Open(
+		sqlite.Open(fmt.Sprintf(ConnectionString, settings.Settings.DB.Path)),
+		&gorm.Config{
+			NamingStrategy: &schema.NamingStrategy{
+				SingularTable: true,
+				NoLowerCase:   true,
+			},
+		})
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	err = db.AutoMigrate(append(model.All())...)
+	if err != nil {
+		err = liberr.Wrap(err)
+		return
+	}
+	return
+}
+
+//
+// Check whether the database has already been migrated to the given version.
+func isMigrated(db *gorm.DB, version string) (migrated bool, err error) {
+	setting := &model.Setting{}
+	result := db.Where("key", api.HubDBVersion).Find(setting)
+	if result.Error != nil {
+		err = result.Error
+		return
+	}
+
+	if setting.Value.String() >= fmt.Sprintf("\"%s\"", version) {
+		migrated = true
+	}
+
+	return
+}
+
+//
+// Perform any necessary DB migrations.
+func migrate(db *gorm.DB) (err error) {
+	var migrations = []func(*gorm.DB) error{
+		v210,
+	}
+
+	db.Exec("PRAGMA foreign_keys = OFF")
```

Comment:
GORM compensates for sqlite not natively supporting changing constraints by creating a new table, copying data, and dropping the old one. We have to shut off the foreign key pragma first, otherwise this will result in data being lost.